Thumbs-up function 👍

mkeeda opened the pull request:
## Issue
- close #429

## Overview (Required)
- Implement thumbs-up button like medium's crap
- Thumbs-up counts are shared to all users
- Only request to update the thumbs-up count if a particular timespan has passed

## Links
-

## Screenshot
Before | After
:--: | :--:
<img src="https://user-images.githubusercontent.com/1386930/72677159-ccb74380-3adc-11ea-8d30-cc13e8b44840.png" width="300" /> | <img src="https://user-images.githubusercontent.com/12047582/74599365-e17afd00-50c3-11ea-8a4e-96ffdf0d1975.gif" width="300" />

takahirom commented:
👀

takahirom commented:
I commented in this 🙏 
https://github.com/DroidKaigi/conference-app-2020/issues/429#issuecomment-578519980

takahirom commented:
I applied some fixes. Can you check it? 🙏 
https://github.com/DroidKaigi/conference-app-2020/commit/4a2de2517c3f9534ca0f1d34283caeecab2fd9c1

takahirom commented:
![image](https://user-images.githubusercontent.com/1386930/73605586-6bff2f00-45e3-11ea-86d2-5dd468c2c0f0.png)

mkeeda commented on data/firestore/src/main/java/io/github/droidkaigi/confsched2020/data/firestore/internal/FirestoreImpl.kt:
```diff
@@ -83,6 +120,40 @@ internal class FirestoreImpl @Inject constructor() : Firestore {
+        try {
+            Tasks.whenAll(tasks).await()
+        } catch(e: Exception) {
+            println("⭐" + e.message)
```
Umm...
I still catch the exception like this.
```
2020-02-02 19:09:35.472 7815-7815/io.github.droidkaigi.confsched2020.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: io.github.droidkaigi.confsched2020.debug, PID: 7815
    com.google.firebase.firestore.FirebaseFirestoreException: PERMISSION_DENIED: Missing or insufficient permissions.
        at com.google.firebase.firestore.util.Util.exceptionFromStatus(com.google.firebase:firebase-firestore@@20.2.0:121)
```

takahirom replied:
👀

mkeeda replied:
Fixed by bf57b49
The cause was `isEmpty`.
https://github.com/DroidKaigi/conference-app-2020/pull/626#issuecomment-581157454

mkeeda commented:
@takahirom 
Thank you for fix my code
I checked your code, but I caught the permission error too.
https://github.com/DroidKaigi/conference-app-2020/pull/626#discussion_r373835072
286d6bd debugging code is used by me.

Now, it is failure to retrieve shards and to create shards.
I cannot access to `thumbsup_counters` collection, probably.
Is my development enviroment some wrong?

I tried the simple code that retrieves documents of `thumbsup_counters`, but I cannot retrieve documents and catch the error.

code
``` kotlin
FirebaseFirestore
    .getInstance()
    .collection("confsched/2020/sessions/155510/thumbsup_counters")
    .get()
    .addOnFailureListener {
        println(it)
    }
    .addOnSuccessListener {
        println(it.documents)
    }
```

<details>
<summary>error</summary>
2020-02-02 19:35:30.056 9913-9913/io.github.droidkaigi.confsched2020.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: io.github.droidkaigi.confsched2020.debug, PID: 9913
    com.google.android.gms.tasks.RuntimeExecutionException: com.google.firebase.firestore.FirebaseFirestoreException: PERMISSION_DENIED: Missing or insufficient permissions.
        at com.google.android.gms.tasks.zzu.getResult(Unknown Source:15)
        at io.github.droidkaigi.confsched2020.data.firestore.internal.FirestoreImpl$getFavoriteSessionIds$2.onComplete(FirestoreImpl.kt:54)
        at com.google.android.gms.tasks.zzj.run(Unknown Source:4)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: com.google.firebase.firestore.FirebaseFirestoreException: PERMISSION_DENIED: Missing or insufficient permissions.
        at com.google.firebase.firestore.util.Util.exceptionFromStatus(com.google.firebase:firebase-firestore@@20.2.0:121)
        at com.google.firebase.firestore.core.EventManager.onError(com.google.firebase:firebase-firestore@@20.2.0:134)
        at com.google.firebase.firestore.core.SyncEngine.handleRejectedListen(com.google.firebase:firebase-firestore@@20.2.0:399)
        at com.google.firebase.firestore.core.FirestoreClient.handleRejectedListen(com.google.firebase:firebase-firestore@@20.2.0:287)
        at com.google.firebase.firestore.remote.RemoteStore.processTargetError(com.google.firebase:firebase-firestore@@20.2.0:555)
        at com.google.firebase.firestore.remote.RemoteStore.handleWatchChange(com.google.firebase:firebase-firestore@@20.2.0:436)
        at com.google.firebase.firestore.remote.RemoteStore.access$100(com.google.firebase:firebase-firestore@@20.2.0:53)
        at com.google.firebase.firestore.remote.RemoteStore$1.onWatchChange(com.google.firebase:firebase-firestore@@20.2.0:176)
        at com.google.firebase.firestore.remote.WatchStream.onNext(com.google.firebase:firebase-firestore@@20.2.0:108)
        at com.google.firebase.firestore.remote.WatchStream.onNext(com.google.firebase:firebase-firestore@@20.2.0:38)
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.lambda$onNext$1(com.google.firebase:firebase-firestore@@20.2.0:119)
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver$$Lambda$2.run(Unknown Source:4)
        at com.google.firebase.firestore.remote.AbstractStream$CloseGuardedRunner.run(com.google.firebase:firebase-firestore@@20.2.0:67)
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.onNext(com.google.firebase:firebase-firestore@@20.2.0:110)
        at com.google.firebase.firestore.remote.FirestoreChannel$1.onMessage(com.google.firebase:firebase-firestore@@20.2.0:120)
        at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:33)
        at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:33)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:563)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at com.google.firebase.firestore.util.AsyncQueue$DelayedStartFactory.run(com.google.firebase:firebase-firestore@@20.2.0:205)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: io.grpc.StatusException: PERMISSION_DENIED: Missing or insufficient permissions.
2020-02-02 19:35:30.056 9913-9913/io.github.droidkaigi.confsched2020.debug E/AndroidRuntime:     at io.grpc.Status.asException(Status.java:541)
        at com.google.firebase.firestore.util.Util.exceptionFromStatus(com.google.firebase:firebase-firestore@@20.2.0:119)
        	... 26 more
</details>

takahirom commented:
We haven't changed any rules since we created the issue.
Perhaps you can create that collection, but you don't have permission to get it, so it is expected that isEmpty will be true.
I am thinking about rule settings that can only get the size.

```
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /confsched/2020 {
...
      match /sessions/{sessionId} {
        match /{allSubcollections=**} {
          allow create
        }
        // Allow to increment only the 'shards' field and only by 1.
        match /thumbsup_counters/{counterId} {
          allow get;
          allow write: if request.resource.data.keys() == ["shards"]
                         && (resource == null || request.resource.data.shards ==
                         resource.data.shards + 1);
        }
      }
    }
  }
}
```

takahirom commented:
I changed the rule 🙏 

```
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /confsched/2020 {
...
      match /sessions/{sessionId} {
        match /{allSubcollections=**} {
          allow create
        }
        match /thumbsup_counters {
          allow read
       	  // Allow to increment only the 'shards' field and only by 1.
          match /{counterId} {
            allow write: if request.resource.data.keys() == ["shards"]
                           && (resource == null || request.resource.data.shards ==
                           resource.data.shards + 1);
          }
        }
      }
    }
  }
}
```

takahirom commented:
I changed!
```
      match /sessions/{sessionId} {
        match /{allSubcollections=**} {
          allow create
        }
        match /thumbsup_counters/{counterId} {
            allow read;
        	  // Allow to increment only the 'shards' field and only by 1.
            allow write: if request.resource.data.keys() == ["shards"]
                           && (resource == null || request.resource.data.shards ==
                           resource.data.shards + 1);
        }
      }
```

takahirom commented on data/firestore/src/main/java/io/github/droidkaigi/confsched2020/data/firestore/internal/FirestoreImpl.kt:
```diff
@@ -83,6 +120,40 @@ internal class FirestoreImpl @Inject constructor() : Firestore {
+    }
+
+    companion object {
+        const val NUM_SHARDS = 10
```
If there seems to be no problem in speed, I want to change it to 5 for cost.

```suggestion
        const val NUM_SHARDS = 5
```

mkeeda replied:
5 shards are able to update 5 times for 1 second.
That means that throughput is 5.

1. 100 persons tap thumbs-up button 10 times for 60 seconds. (very popular session?)
Requirement throughput
100 * 10 / 60 ≒ 16.7

1. 100 persons tap thumbs-up button 5 times for 60 seconds. (busy?)
Requirement throughput
100 * 5 / 60 ≒ 8.33

1. 50 persons tap thumbs-up button 5 times for 60 seconds. (normaly?)
Requirement throughput
50 * 5 / 60 ≒ 4.16

Umm... 
I think enough speed, but I do not have confidence😅

takahirom replied:
If firestore is free, I did so, but I am concerned that the price will be expensive because every time the value flows, the number of reads runs to everyone.

mkeeda replied:
I forgot firebase pricing.
I have to reduce accesses of firestore.
I'm thinking it a little more.

takahirom replied:
Please tell me your opinion 🙏 I think we can thumbs up 300 / minutes. Is it not enough?

takahirom replied:
Let's calculate a little.
100 people are watching * 10 shards * (100 people * 100 times) * 40 sessions = 400 million read
Read Document $ 0.038 per 100,000 Documents
0.038 * (100 * 10 * 100 * 100 * 40) / 100000
= $ 152.0

(100 people * 100 times) * 40 sessions = 400000
Write Document $ 0.115 per 100,000 Documents

0.115 * 100*100*40 / 100000
= $ 0.46

About 15,000 yen (  Organizer @mhidaka says it is OK)

But you may be able to do something. 👀 
For example, if you have 9 shards and monitor only 3 shards, you can multiply by 3 to get 1/3.

takahirom replied:
Even if this does not reflect your thumbs-up, it may not be good.

mkeeda replied:
I have ideas for reduce access to firestore.
So I think 5 shards is good.
https://github.com/DroidKaigi/conference-app-2020/issues/429#issuecomment-582482685

I apply your change!

mkeeda commented:
The code still fail to create shards...
I will find out the cause tomorrow.

takahirom commented:
Maybe even if there is a value, I think that isEmpty is true.
Personally, I think it's a good idea to check if there are 9 counters so that the process can die during creation.
Or it might be better to use something like a transaction
https://firebase.google.com/docs/firestore/manage-data/transactions

takahirom commented on corecomponent/androidcomponent/src/main/res/values/themes.xml:
```diff
@@ -10,6 +10,7 @@
         <item name="textAppearanceBody1">@style/TextAppearance.DroidKaigi.Body1</item>
         <item name="textAppearanceBody2">@style/TextAppearance.DroidKaigi.Body2</item>
         <item name="textAppearanceCaption">@style/TextAppearance.DroidKaigi.Caption</item>
+        <item name="textAppearanceButton">@style/TextAppearance.DroidKaigi.Button</item>
```
👍

takahirom commented:
I changed the rule 🙏 

```
                           request.resource.data.shards > resource.data.shards &&
                           request.resource.data.shards < resource.data.shards + 51
```

mkeeda commented on feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/viewmodel/SessionDetailViewModel.kt:
```diff
@@ -76,18 +94,30 @@ class SessionDetailViewModel @AssistedInject constructor(
                 ?: favoriteState
+                    .getErrorIfExists()
+                    .toAppError()
+                ?: thumbsUpCountLoadState
```
What is better code?

mkeeda commented:
## TODO
- [ ] Fix tests
- [ ] Implement error handling when increment thumbs-up count is failure

takahirom commented:
I implemented debounce logic 🙏
https://github.com/DroidKaigi/conference-app-2020/commit/be331b0421f750d8ade541c5c1be3d3b36ef8dd5

mkeeda commented on data/firestore/src/main/java/io/github/droidkaigi/confsched2020/data/firestore/internal/FirestoreImpl.kt:
```diff
@@ -76,9 +87,41 @@ internal class FirestoreImpl @Inject constructor() : Firestore {
+            .debounce(INCREMENT_DEBOUNCE_MILLIS)
+            .map {
+                val result = minOf(it.index - appliedCount, MAX_APPLY_COUNT)
+                appliedCount = it.index
```
`appliedCount` is incremented unlimited?

takahirom replied:
Yes
appliedCount exists for it.index - appliedCount
it.index is incremented unlimited. So appliedCount is incremented unlimited too.

mkeeda replied:
Thank you, I understood.
`appliedCount` is a index of `thumbsUpEventChannel`that is already posted to firestore.
And `appliedCount`is a thumbs up count to should be incremented too.

mkeeda commented:
@takahirom 
I moved the debouce processing ViewModel, and removed the code of checking index.
Probably, the debounce processing is a bussiness logic, so`Firestore` should not have it.
67e699a

And I want to show own incrementd count like medium.
So I separated increment flow from total thumbs up count flow.
https://github.com/DroidKaigi/conference-app-2020/issues/429#issuecomment-582482685

takahirom commented:
:eyes:

takahirom commented:
As for business logic, it is difficult because there is a possibility that duplicate logic will be created by bringing it to the ViewModel. However, if you want to have incrementThumbsUpCount in UiModel, I think it is good.

takahirom commented on data/firestore/src/test/java/io/github/droidkaigi/confsched2020/data/firestore/internal/FirestoreImplTest.kt:
```diff
@@ -0,0 +1,59 @@
+import org.junit.Test
+
+class FirestoreImplTest {
+    @Ignore
```
Sorry please remove this :bow:

takahirom commented on feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/viewmodel/SessionDetailViewModel.kt:
```diff
@@ -107,11 +165,44 @@ class SessionDetailViewModel @AssistedInject constructor(
+                    count = count
+                )
+                Timber.debug { "⭐ increment $count posted" }
+                incrementThumbsUpCountLiveData.value = 0
```
If you add the totalThumbsUpCount and the incrementThumbsUpCount to display, please note that some logic such as scan is required.
Perhaps if you put 0 in this timing, if you start with 10 thumbs-up and hit thumbs-up 5 times, it will instantly become 20thumbs-up.
If you don't add them, it's OK.

takahirom commented:
For now, let's proceed with the implementation of displaying incrementThumbsUpCount! 👍

takahirom commented:
@mkeeda Any progress...? 🙇

mkeeda commented:
@takahirom 
Sorry, I made you wait 🙇 🙇 🙇
I implemented popup animation!
![thumbsup](https://user-images.githubusercontent.com/12047582/74446038-ab1e6180-4eba-11ea-96d5-875aa7729c31.gif)

### TODO
- [ ] Fix tests
- [ ] Implement error handling when increment thumbs-up count is failure

takahirom commented:
👀

mkeeda commented:
The popup animation is too moved?
It is little hard to watch a incremented count for me.

takahirom commented on corecomponent/androidcomponent/src/main/java/io/github/droidkaigi/confsched2020/ext/Animators.kt:
```diff
@@ -0,0 +1,41 @@
+
+// This function is copied from https://medium.com/androiddevelopers/suspending-over-views-19de9ebd7020
+
+suspend fun Animator.awaitEnd() = suspendCancellableCoroutine<Unit> { cont ->
```
👍

takahirom commented on feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/viewmodel/SessionDetailViewModel.kt:
```diff
@@ -76,21 +109,49 @@ class SessionDetailViewModel @AssistedInject constructor(
         )
     }
 
+    val appError: LiveData<AppError?> = merge(
```
I have a question. Please tell me why you separated the error 🙋‍♂ 
for simplifying??

mkeeda replied:
Yes.
The reason is that uiModel's errors is complexed.
In my opinion, it is easy to read that the code is separated success and failure flow.

What do you think?

takahirom replied:
You have added the errorGettable interface. 👍 
I think that you can write enough with UiModel by using this. 👀 
https://github.com/DroidKaigi/conference-app-2020/commit/70f7d25873760918b339a90d659e7f30e58db55c

takahirom replied:
Also, besides whether it is good or bad, I would like to challenge this DroidKaigi to make one UiModel.

mkeeda replied:
OK, let's challenge the one UiModel!
70f7d25 is enough to simple too for me.

takahirom commented:
@mkeeda 
Thanks as always 🙏 
Can you take time today? 
What are the remaining tasks now?

mkeeda commented:
@takahirom 
Sorry, I’m almost done!
I will finish tasks today.
The remaining tasks ↓
- [ ] Cherry-pick 70f7d25 (Conflicting my changes...)
- [ ] Refactor animation functions
- [ ] Remove debug log (I want to remove ⭐print)

I will write overviews at top of this PR when I finished all tasks.

takahirom commented:
Thanks! I want to merge this today!! 👍

takahirom commented:
Can you delete WIP? 😄

mkeeda commented:
Yes!
I'm writing overviews now!

mkeeda commented:
@takahirom 
All tasks is done!
Please review again 🙏 🙏 🙏

takahirom commented:
LGTM 👍👍👍👍
